cmd/builder/internal/builder/config.go

```diff
@@ -207,34 +203,9 @@ func (c *Config) ParseModules() error {
 		return err
 	}
 
-	if c.Providers != nil {
-		providers, err := parseModules(*c.Providers)
-		if err != nil {
-			return err
-		}
-		c.Providers = &providers
-	} else {
-		providers, err := parseModules([]Module{
-			{
-				GoMod: "go.opentelemetry.io/collector/confmap/provider/envprovider v" + c.Distribution.OtelColVersion,
-			},
-			{
-				GoMod: "go.opentelemetry.io/collector/confmap/provider/fileprovider v" + c.Distribution.OtelColVersion,
-			},
-			{
-				GoMod: "go.opentelemetry.io/collector/confmap/provider/httpprovider v" + c.Distribution.OtelColVersion,
-			},
-			{
-				GoMod: "go.opentelemetry.io/collector/confmap/provider/httpsprovider v" + c.Distribution.OtelColVersion,
-			},
-			{
-				GoMod: "go.opentelemetry.io/collector/confmap/provider/yamlprovider v" + c.Distribution.OtelColVersion,
-			},
-		})
-		if err != nil {
-			return err
-		}
-		c.Providers = &providers
+	c.Providers, err = parseModules(c.Providers)
```

Comment:
I fear this will break most ocb users. If we go this route, is there a gentler approach we could take with feature gates or timelines?

Reply:
>The main problem with this, is that the providers will start to have a different version than the builder and they will be independently mark as stable which will make it very hard to track all combination of builder versions with the "associated default" providers for that version.

> This also fixes an issue, that for builder 0.109.0 the default providers will not work since the env and file are no longer using 0.109.0 version.

Reply:
I would have suggested to actually wait for all providers to be marked stabled in one version, that would have simplified a lot. Otherwise, I am not sure.

Reply:
>I fear this will break most ocb users. If we go this route, is there a gentler approach we could take with feature gates or timelines? 

Also, right now they are broke anyway. Because if they don't specify any provider, it will try to use env/file at 109 and will fail.

Reply:
I don't think there's much that prevents us from marking the other providers as 1.0, I will open the PR for those

Reply:
I don’t think we should rush the stability.

Reply:
@TylerHelmuth revisiting this, if my memory serves we have had two issues about this in the last three weeks, so I am tempted to say we should just do this breaking change since it hasn't been a huge deal for people. Do you think we should wait more/do it differently?

Reply:
Will doing this mean anyone using ocb who doesn't include providers in the manifests.yaml will build a collector with 0 providers? I am pretty sure a collector setup like that will compile but not run, as we require at least 1 provider.

If that is true, then users who don't specify providers (most I'm guessing) will start getting runtime errors with their collectors. I much rather we do this gracefully instead of one breaking change.

If the assumptions I've made are wrong then the risk goes down.

Reply:
I'd like to see a confirmation on what happens with this change if a manifest is fed to ocb with no providers configured.

Reply:
I made it clear that ocb command will fail with a clear message. So customer will not get into production bad collectors that are unable to start.

Reply:
I'd personally rather see this as a incremental change behind a feature gate instead of a breaking change.

We do incremental changes behind feature gates for breaking changes that only affect component developers and library authors, I think the same pattern should apply to end-user breaking changes.

Reply:
I am fine with that, although we need to add support for feature gates on ocb first then :smile:

Reply:
😿

Reply:
If feature gates support in ocb isn't digestible, then we should tell users who experience the breaking change exactly what to put in their manifest to get back to the previous behavior

Reply:
I am fine with adding support, also fine with just a temporary CLI flag for this that we eventually remove

Reply:
For this fix, having a flag or changing minimal the config I am confused where the benefit is. Also to remember this was and is broken.

Reply:
I am 100% to add support in the future for featureflags.

Reply:
For this particular change, I am fine with any of
- an error message that tells you what do very explicitly
- a feature gate
- a dedicated config flag

The "at least one provider is required" may not be super clear to people (they may not be aware of what providers they used to put in there).

Reply:
Agreed.  Since this breaking change will affect end users of OCB who were not already setting providers, they'll be the most confused about what a provider is and what they need to set.  We need to make it obvious what to do from the error message and the changelog entry (and I believe the OCB readme as well), so that they can learn how to fix their manifest.